pkg/crc/api/api_http.go

```diff
@@ -0,0 +1,156 @@
+)
+
+var handler *Handler
+
+func NewMux(config crcConfig.Storage, machine machine.Client) http.Handler {
+	handler = NewHandler(config, machine)
+
+	mux := http.NewServeMux()
+	mux.HandleFunc("/start", func(w http.ResponseWriter, r *http.Request) {
+		var startResult string
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			logging.Debugf("Unable to read request body for URI: %s", r.RequestURI)
+			http.Error(w, "Unable to read request body", http.StatusInternalServerError)
+			return
+		}
+		if len(data) > 0 {
+			startResult = handler.Start(data)
+		} else {
+			startResult = handler.Start(nil)
+		}
+		sendResponse(w, startResult)
+	})
+
+	mux.HandleFunc("/stop", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
+			return
+		}
+		stopResult := handler.Stop()
+		sendResponse(w, stopResult)
+	})
+
+	mux.HandleFunc("/status", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
+			return
+		}
+		status := handler.Status()
+		sendResponse(w, status)
+	})
+
+	mux.HandleFunc("/delete", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
+			return
+		}
+		deleteResult := handler.Delete()
+		sendResponse(w, deleteResult)
+	})
+
+	mux.HandleFunc("/version", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
+			return
+		}
+		versionResult := handler.GetVersion()
+		sendResponse(w, versionResult)
+	})
+
+	mux.HandleFunc("/webconsoleurl", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodGet) {
+			return
+		}
+		webconsoleInfo := handler.GetWebconsoleInfo()
+		sendResponse(w, webconsoleInfo)
+	})
+
+	mux.Handle("/config/", http.StripPrefix("/config", newConfigMux(handler)))
+
+	return mux
+}
+
+func newConfigMux(handler *Handler) http.Handler {
+	mux := http.NewServeMux()
+
+	mux.HandleFunc("/get", func(w http.ResponseWriter, r *http.Request) {
+		var configGetResult string
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			logging.Debugf("Unable to read request body for URI: %s", r.RequestURI)
+			http.Error(w, "Unable to read request body", http.StatusInternalServerError)
+			return
+		}
+		if len(data) > 0 {
+			configGetResult = handler.GetConfig(data)
+		} else {
+			configGetResult = handler.GetConfig(nil)
+		}
+		sendResponse(w, configGetResult)
+	})
+
+	mux.HandleFunc("/set", func(w http.ResponseWriter, r *http.Request) {
+		if wrongHTTPMethodUsed(r, w, http.MethodPost) {
+			return
+		}
+
+		data, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			logging.Debugf("Unable to read request body for URI: %s", r.RequestURI)
+			http.Error(w, "Unable to read request body", http.StatusInternalServerError)
+			return
+		}
+		if len(data) < 1 {
+			http.Error(w, "Not enough arguments for /config/set", http.StatusBadRequest)
+			return
+		}
+		configSetResult := handler.SetConfig(data)
```

Comment:
Fwiw, that's a lot of copy&pasted code when the only difference between each block of code is this line.

Reply:
Changed, not ideal but reduced the duplicating lines a bit.